### DISCUSSION TITLE
Image Drawing Module

### DISCUSSION THREAD

**Discussion by RobLoach** [raysan5/raylib#2865] [Features]
Been wanting to use raylib's Image API outside of raylib. One use case is in a [libretro](https://www.libretro.com/) core that would allow using parts of raylib on [devices that Lakka supports](http://lakka.tv/).  This would allow compiling once, and rendering in multiple different frontends that use libretro, like RetroArch.

## Proof of Concept

I forked raylib's [rtextures.c](https://github.com/raysan5/raylib/blob/8e57c8ace389135d1f4b92ced0bfb79920ed359e/src/rtextures.c) to its own single-header library over at [rimage](https://github.com/robloach/rimage). Following that, I created [libretro-rimage](https://github.com/robloach/libretro-rimage), which allows you to use the Image API in a libretro core. Using the framebuffer and raylib's Image API makes it so much easier to render things to the screen.

``` c
ImageDrawCircle(&core->backBuffer, 300, 100, 30, SKYBLUE);
ImageDrawRectangle(&core->backBuffer, 100, 100, 100, 140, GREEN);
ImageDrawLine(&core->backBuffer, 50, 50, 200, 80, ORANGE);
ImageDrawRectangle(&core->backBuffer, 20, 150, 50, 80, RED);
```

![Screenshot at 2023-01-17 12-38-01](https://user-images.githubusercontent.com/25086/212971874-d7c3e65d-de28-48ed-87ea-df2c76b3038c.png)

This fork is a big hack, and I would like a cleaner design around it so that a full fork isn't needed.

## Proposed Solution

I would love to move the Image API that's done in `rtextures.c` into its own single-header library, similar to what was done with [`rl_gputex.h`](https://github.com/raysan5/raylib/blob/178a356cb4130f942f051f5669f46ebcb7f438bc/src/external/rl_gputex.h). This would allow using raylib's Image API outside of raylib, without requiring a complete fork.

## Challenges

There are a few problems that we can run into when taking this approach....

- Duplicate function definitions
- Duplicate type names
- Having to sync things across this new single-header file and raylib.h
- Dependency issues, like having the Font stuff available to the Image API

Would love all of your thoughts. Introducing more modularity to raylib is great, but I don't want to over-complicate the code base.

**Comment by RobLoach** [comment]
Perhaps to avoid some of the duplicate definition problems, we have a similar approach that was done with `rl_gputex.h` with function names, avoid the raylib structs, and just have the raylib API call the library's functions...

``` c
void rl_imagedrawpixel(void* data, int format, int width, int height,
        int posX, int posY,
        unsigned char colorR,
        unsigned char colorG,
        unsigned char colorB,
        unsigned char colorA);

void ImageDrawPixel(Image* dst, int posX, int posY, Color color) {
    rl_imagedrawpixel(dst->data, dst->format, dst->width, dst->height,
        posX, posY,
        color.r, color.g, color.b, color.a);
}
```

Avoiding the structs could make the function definitions extremely long in the library, but the benefit of not having to deal with the duplicate structs could be worth it.

**Comment by JeffM2501** [comment]
I'm not sure I'd call it a software rendering module. I'd call it **Image Drawing**, Software rendering makes it sound like it's a full blown 3d renderer on par with OpenGL, and it is not.

**Reply by RobLoach** [reply]
Good call, renamed the discussion. Definitely more image manipulation than rendering.

**Reply by raysan5** [reply]
Actually SDL calls the internal software renderer that same name despite only supporting 2d rendering...

**Comment by konsumer** [comment]
I think this idea is great. Drawing stuff on the screen is a bit complicated in libretro without this, but very familiar & easy with it.

**Comment by raysan5** [comment]
@RobLoach Definitely I want to do something on that line for a 2d-software-rendered or image-drawing functionality but I'm still thinking the best approach for it. My first idea was just keeping `rtextures.c` and enable pure image-rendeing with a flag, allowing the texture functions to fall-back to the image functions (https://github.com/raysan5/raylib/issues/1370) but it increases the complexity of the module a lot, and it won't be a simple *learning-material* anymore for new users.

Moving the image functionality to a separate `rimage.h` it's probably the best approach but then we face the function duplicate issues and the coupling with raylib functionality, and again the complexity of the module is increased (probably not much for programmers with some experience but probably enough for novice programmers to avoid jumping and modifying it...)

I need to think about the best solution but definetly I want to allow it; I've got a side project to run raylib on RP2040 microcontroller (no-GPU, 264KB RAM) and I need the 2d-software-rendering option.

**Reply by RobLoach** [reply]
> raylib on RP2040 microcontroller (no-GPU, 264KB RAM)

:exploding_head: That's pretty awesome! Would be so cool to play some raylib games on a small handheld.

**Reply by raysan5** [reply]
It's on my TODO list, actually I started working on it about one year ago but too many things got in the middle this last year...

**Comment by RobLoach** [comment]
We have a software renderer in raylib now, so I'm going to close this one 👍.... Also, pntr is an option if you want a stripped down version.

**Reply by raysan5** [reply]
Despite we already have a full 3d-pipeline software renderer, I'm still playing with the idea of a _simple 2d software renderer_ based on the `Image` API, one of the previous main stoppers was that no platform backend supported a raw pixels framebuffer window... but that changed with the new 3d software renderer. I think a new revisit of this feature can be considered.

Just playing with some approaches for now, a simple and not-much-intrusive idea is a `raylib_soft2d.h` include that remaps some API functions:
```c
// Framebuffer is an image
static Image *fb;

#define Texture2D           Image
#define RenderTexture2D     Image

/// core.c
// NOTE: Some functions need to be reimplemnted
#define InitWindow      InitWindowSoft
#define ClearBackground(color)  ImageClearBackground(fb, color)

// rshapes
#define DrawPixel(x, y, col)                ImageDrawPixel(fb, x, y, col)
#define DrawPixelV(pos, col)                ImageDrawPixelV(fb, pos, col)
#define DrawLine(x1, y1, x2, y2, col)       ImageDrawLine(fb, x1, y1, x2, y2, col)
#define DrawLineV(start, end, col)          ImageDrawLineV(fb, start, end, col)
#define DrawLineEx(start, end, thick, col)  ImageDrawLineEx(fb, start, end, thick, col)

#define DrawTriangle(v1, v2, v3, col)       ImageDrawTriangle(fb, v1, v2, v3, col)
#define DrawTriangleLines(v1, v2, v3, col)  ImageDrawTriangleLines(fb, v1, v2, v3, col)
#define DrawTriangleFan(points, count, col) ImageDrawTriangleFan(fb, points, count, col)
#define DrawTriangleStrip(points, count, col) ImageDrawTriangleStrip(fb, points, count, col)

#define DrawRectangle(x, y, w, h, col)      ImageDrawRectangle(fb, x, y, w, h, col)
#define DrawRectangleV(pos, size, col)      ImageDrawRectangleV(fb, pos, size, col)
#define DrawRectangleRec(rec, col)          ImageDrawRectangleRec(fb, rec, col)
#define DrawRectangleLines(x, y, w, h, col) ImageDrawRectangleLines(fb, (Rectangle){ x, y, w, h }, 1, col)

#define DrawCircle(x, y, radius, col)       ImageDrawCircle(fb, x, y, radius, col)
#define DrawCircleV(center, radius, col)    ImageDrawCircleV(fb, center, radius, col)
#define DrawCircleLines(x, y, radius, col)  ImageDrawCircleLines(fb, x, y, radius, col)
#define DrawCircleLinesV(center, radius, col) ImageDrawCircleLinesV(fb, center, radius, col)

// rtextures
#define LoadTexture(fileName)               LoadImage(fileName)
#define LoadTextureFromImage(image)         ImageCopy(image)
#define LoadRenderTexture(width, height)    GenImageColor(width, height, BLANK)
#define UnloadTexture(tex)                  UnloadImage(tex)
#define UnloadRenderTexture(target)         UnloadImage(target)
// TODO: Image drawing functions neeed a redesign

// rtext
#define DrawText(text, x, y, size, color)   ImageDrawText(fb, text, x, y, size, color)
#define DrawTextEx(font, text, position, size, spacing, tint) ImageDrawTextEx(fb, font, text, position, size, spacing, tint)
```

What could be the utility for this small library? Higher performance on very low-end embedded devices that just need some 2d drawing.

**Reply by RobLoach** [reply]
Great ideas. Would work for embedded devices. Could likely implement a wrapper to write directly to the FB https://lvntky.github.io/fbgl/

My initial motivation was as a crude libretro core, though there are better means now that I'd like to try out, especially with the learnings from https://github.com/RobLoach/raylib-libretro
